Replace createVehicleCrew with fn_createCrew

PiG13BR opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no <!-- don't forget to update CHANGELOG.md file --> |
| Needs wipe? | no <!-- set to yes if save needs to be wiped to use new feature --> |

### Description:
- Ditch createVehicleCrew command and replace it with a function that uses preset configuration only. Now you can put a different faction assets in the same preset without spawning crew that not fit into your scenario.
- The function checks what kind of vehicle is to get the correct type of crew.
- The function tries to find empty seats for driver, gunner and commander. For the turret seat (generally from helicopters), only gets the co-pilot.

### Content:
- [ x] Add new function called fn_createCrew
- [ x] Replace all createVehicleCrew command in the files
- [ ] Put the new variables (used in the function) in other presets

<!--
Add things which are part of this pull request as checkboxes
to show if it's already finished and already part of the pull request.
-->

### Successfully tested on:
- [ x] Local MP
- [ x] Dedicated MP

<!--
As soon as you've tested your feature on the listed environment you can check the checkbox.
It has to work without any issues and errors in your own tests.
-->

PiG13BR committed:
Update fn_tutorial.fsm

PiG13BR committed:
Update fn_tutorial.fsm

PiG13BR committed:
Update open_arsenal.sqf

PiG13BR committed:
Merge branch 'KillahPotatoes:v0.96.8' into v0.96.8

PiG13BR committed:
Add files via upload

PiG13BR committed:
Update fn_forceBluforCrew.sqf

Add fn_createCrew

PiG13BR committed:
Update CfgFunctions.hpp

Add fn_createCrew

PiG13BR committed:
Update fn_spawnVehicle.sqf

Add fn_createCrew

PiG13BR committed:
Update spawn_air.sqf

Add fn_createCrew

PiG13BR committed:
Update send_paratroopers.sqf

PiG13BR committed:
Update fn_spawnVehicle.sqf

PiG13BR committed:
Update custom.sqf

PiG13BR committed:
Update custom.sqf

some-evil-kitty commented:
Most presets don't have pilots and crew defined yet - maybe add some logic inside the function to default to something if they're not defined? I was thinking rifleman for crew at the very least.

PiG13BR commented:
Yes, I can do that for now, but I could also add those variables on the other presets aswell, it's boring but I can do it with time. Tell me if this is worth it so I can manage my time to add them.

PiG13BR committed:
Placeholder types in fn_createCrew.sqf

PiG13BR commented:
Done. I've put the placeholder types of crew that already exist on presets for each side.

PiG13BR committed:
(0.98) Hotfix arsenal init for CBA_A3 3.18

veteran29 commented on Missionframework/functions/fn_createCrew.sqf:
```diff
@@ -0,0 +1,130 @@
+	if (isNull _checkSeat) then {
+
+		switch (_role) do {
+			case "driver" : {
+				_crew = [_typeCrew, getPos _vehicle, _grp] call KPLIB_fnc_createManagedUnit;
+				_crew assignAsDriver _vehicle;
+				_crew moveInDriver _vehicle;
+			};
+			case "gunner" : {
+				_crew = [_typeCrew, getPos _vehicle, _grp] call KPLIB_fnc_createManagedUnit;
+				_crew assignAsGunner _vehicle;
+				_crew moveInGunner _vehicle;
+			};
+			case "commander" : {
+				_crew = [_typeCrew, getPos _vehicle, _grp] call KPLIB_fnc_createManagedUnit;
+				_crew assignAsCommander _vehicle;
+				_crew moveInCommander _vehicle;
+			};
+			case "turret" : {
+				//Cargo index -1 = copilot (vanilla models)
+				if ((_vehicle isKindOf "Air") && (_index == -1)) then {
+					_crew = [_typeCrew, getPos _vehicle, _grp] call KPLIB_fnc_createManagedUnit;
+					_crew assignAsTurret [_vehicle, _turretPath];
+					_crew moveInTurret [_vehicle, _turretPath];
+				}
+			}
+		};
+
+		// Check if the vehicle crew spawned is in the vehicle, if not, delete it
+		if !(_crew in (crew _vehicle)) then {
+			deleteVehicle _crew;
+		} else {
+			_allCrew pushBack _crew;
+			_crew addMPEventHandler ["MPKilled", {
+				params ["_unit", "_killer"];
+				["KPLIB_manageKills", [_unit, _killer]] call CBA_fnc_localEvent;
+			}];
+		}
+	}
+}forEach _seats;
```
This can be greatly simplified with `BIS_fnc_vehicleCrewTurrets`, `createManagedUnit` aleady takes care about adding the  kill manager.

```suggestion
private _grp = createGroup [_side, true];
private _allCrew = [];

{
	private _crew = [_typeCrew, getPosATL _vehicle, _grp] call KPLIB_fnc_createManagedUnit;
	_allCrew pushBack _crew;

	if (_x isEqualTo [-1]) then {
		_crew assignAsDriver _vehicle;
		_crew moveInDriver _vehicle;
	} else {
		_crew assignAsTurret [_vehicle, _x];
		_crew moveInTurret [_vehicle, _x];
	};
} forEach (_vehicle call BIS_fnc_vehicleCrewTurrets);

```

PiG13BR replied:
Wow nice catch! In my searchs about turrets this function never came up.

veteran29 commented on Missionframework/functions/fn_createCrew.sqf:
```diff
@@ -0,0 +1,130 @@
+
+params[
+	["_vehicle", objNull, [objNull]],
+	["_side", KPLIB_side_enemy,[sideUnknown]]
```
```suggestion
	["_side", KPLIB_side_enemy, [sideUnknown]]

```

veteran29 commented on Missionframework/functions/fn_createCrew.sqf:
```diff
@@ -0,0 +1,130 @@
+}forEach _seats;
+
+// Return all the vehicle crew and its group
+[_allCrew,_grp]
```
```suggestion
[_allCrew, _grp]

```

PiG13BR committed:
Update Missionframework/functions/fn_createCrew.sqf

- Add BIS_fnc_vehicleCrewTurrets to simplify the script.
- Removed the MPKilled event (createManagedUnit already add this event)

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>

PiG13BR committed:
Update Missionframework/functions/fn_createCrew.sqf

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>

PiG13BR committed:
Update Missionframework/functions/fn_createCrew.sqf

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>

some-evil-kitty commented on Missionframework/scripts/client/misc/fn_initArsenal.sqf:
Could we nix this change? It's out of scope and covered by #970

PiG13BR committed:
Merge branch 'KillahPotatoes:v0.96.8' into v0.96.8
